payjoin-cli/src/app/v2.rs

```diff
@@ -45,63 +49,82 @@ impl AppTrait for App {
+        let url = uri.extras.endpoint();
+        // match bip21 to send_session public_key
+        let req_ctx = match self.db.get_send_session(url)? {
+            Some(send_session) => send_session,
+            None => {
+                let mut req_ctx = self.create_pj_request(&uri, fee_rate)?;
+                self.db.insert_send_session(&mut req_ctx, url)?;
+                req_ctx
+            }
         };
-        log::debug!("Awaiting response");
-        let res = self.long_poll_post(&mut req_ctx).await?;
-        self.process_pj_response(res)?;
-        self.db.clear_send_session()?;
-        Ok(())
+        self.spawn_payjoin_sender(req_ctx).await
     }
 
-    async fn receive_payjoin(self, amount_arg: &str, is_retry: bool) -> Result<()> {
-        use payjoin::receive::v2::Enroller;
+    async fn receive_payjoin(self, amount_arg: &str) -> Result<()> {
+        use payjoin::receive::v2::SessionInitializer;
 
+        let address = self.bitcoind()?.get_new_address(None, None)?.assume_checked();
+        let amount = Amount::from_sat(amount_arg.parse()?);
         let ohttp_keys = unwrap_ohttp_keys_or_else_fetch(&self.config).await?;
-        let mut enrolled = if !is_retry {
-            let mut enroller = Enroller::from_directory_config(
-                self.config.pj_directory.clone(),
-                ohttp_keys.clone(),
-                self.config.ohttp_relay.clone(),
-            );
-            let (req, ctx) =
-                enroller.extract_req().map_err(|e| anyhow!("Failed to extract request {}", e))?;
-            println!("Starting new Payjoin session with {}", self.config.pj_directory);
-            let http = http_agent()?;
-            let ohttp_response = http
-                .post(req.url)
-                .header("Content-Type", payjoin::V2_REQ_CONTENT_TYPE)
-                .body(req.body)
-                .send()
-                .await
-                .map_err(map_reqwest_err)?;
+        let mut initializer = SessionInitializer::new(
+            address,
+            self.config.pj_directory.clone(),
+            ohttp_keys.clone(),
+            self.config.ohttp_relay.clone(),
+            std::time::Duration::from_secs(60 * 60),
+        );
+        let (req, ctx) =
+            initializer.extract_req().map_err(|e| anyhow!("Failed to extract request {}", e))?;
+        println!("Starting new Payjoin session with {}", self.config.pj_directory);
+        let http = http_agent()?;
+        let ohttp_response = http
+            .post(req.url)
+            .header("Content-Type", payjoin::V2_REQ_CONTENT_TYPE)
+            .body(req.body)
+            .send()
+            .await
+            .map_err(map_reqwest_err)?;
+        log::error!("{:?}", ohttp_response.status());
+        let session = initializer
+            .process_res(ohttp_response.bytes().await?.to_vec().as_slice(), ctx)
+            .map_err(|e| anyhow!("Enrollment failed {}", e))?;
+        self.db.insert_recv_session(session.clone())?;
```

Comment:
Should it follow the same "get or create session" pattern used in `send_payjoin` here?

Reply:
I didn't have expectations one way or another but was wondering about the different treatment for send vs. receive. This makes sense to me. 

> Perhaps receive should instruct one to use resume when the signal to shut down is given.

Yeah, that sounds nice (same with send if it's still waiting on a response when shut down).